reed-solomon-novelpoly/src/field/ops.rs

```diff
@@ -0,0 +1,101 @@
+
+use super::{FieldT, Castomat};
+
+use derive_more::{Add, AddAssign, BitXor, BitXorAssign, Sub, SubAssign};
+
+/// Additive via XOR form of f2e16
+#[derive(Clone, Copy, Debug, Default, BitXor, BitXorAssign, PartialEq, Eq)] // PartialOrd,Ord
+pub struct Additive<F: FieldT>(pub F::Element);
+
+impl<F: FieldT> Additive<F> {
+    #[inline(always)]
+	pub fn to_wide(self) -> F::Wide {
+		F::Wide::from(self.0)
+	}
+
+    #[inline(always)]
+	pub fn from_wide(x: F::Wide) -> Additive<F> {
+		Additive(F::Element::from(x))
+	}
+
+	const ZERO: Additive<F> = Additive(0.cast_as() as F::Element);
+}
+
+#[cfg(table_bootstrap_complete)]
+impl<F: FieldT> Additive<F> {
+	/// Return multiplier prepared form
+    #[inline(always)]
+	pub fn to_multiplier(self) -> Multiplier<F> {
+		Multiplier(LOG_TABLE[self.0.cast_as()])
+	}
+
+	/// Return a*EXP_TABLE[b] over GF(2^r)
+    #[inline(always)]
+	pub fn mul(self, other: Multiplier<F>) -> Additive<F> {
+		if self == Self::ZERO {
+			return Self::ZERO;
+		}
+		let log = F::Wide::cast_from(LOG_TABLE[self.0.cast_as()]) + F::Wide::cast_from(other.0);
+		let offset = F::Wide::cast_from(log & ONEMASK) + (log >> FIELD_BITS);
+		Additive(EXP_TABLE[offset.cast_as()])
+	}
+
+	/// Multiply field elements by a single multiplier, using SIMD if available
+    #[inline(always)]
+	pub fn mul_assign_slice(selfy: &mut [Self], other: Multiplier<F>) {
+		for s in selfy {
+			*s = s.mul(other);
+		}
+	}
+}
+
+
+/// Multiplicaiton friendly LOG form of f2e16
+#[derive(Clone, Copy, Debug, Add, AddAssign, Sub, SubAssign, PartialEq, Eq)] // Default, PartialOrd,Ord
+pub struct Multiplier<F: FieldT>(pub F::Element);
```

Comment:
This does not work.  Multiplier is necessarily an associated type.

Reply:
It does. The implementation can be specific to `F`.

Reply:
No.  I've told you repeatedly this code shall become more complex.  I've considered and rejected exactly this trait and many smaller variations upon it.  The approach in #11 avoids back tracking.